mito-ai/src/utils/checkpoint.ts

```diff
@@ -0,0 +1,188 @@
+            cellCheckpoint.outputs = codeModel.outputs ? codeModel.outputs.toJSON() : [];
+            cellCheckpoint.executionCount = codeModel.executionCount;
+        }
+
+        cells.push(cellCheckpoint);
+    });
+
+    const checkpoint: INotebookCheckpoint = {
+        checkpointId,
+        timestamp: new Date(),
+        notebookPath: currentWidget.context.path || null,
+        cells,
+        metadata: notebook.model?.metadata || {}
+    };
+
+    return checkpoint;
+};
+
+/**
+ * Generates a unique checkpoint ID
+ */
+const generateCheckpointId = (): string => {
+    return `checkpoint_${Date.now()}_${Math.random().toString(36).substr(2, 9)}`;
+};
+
+// Simple in-memory storage for the current session
+let currentCheckpoint: INotebookCheckpoint | null = null;
+
+/**
+ * Saves a checkpoint to storage (currently in-memory for simplicity)
+ */
+export const saveCheckpoint = (checkpoint: INotebookCheckpoint): void => {
+    currentCheckpoint = checkpoint;
+    console.log(`💾 Checkpoint saved: ${checkpoint.checkpointId}`);
+};
+
+/**
+ * Gets the currently stored checkpoint
+ */
+export const getCurrentCheckpoint = (): INotebookCheckpoint | null => {
+    return currentCheckpoint;
+};
+
+/**
+ * Restores the notebook to the state stored in the checkpoint
+ */
+export const restoreFromCheckpoint = async (notebookTracker: INotebookTracker, checkpoint: INotebookCheckpoint): Promise<boolean> => {
+    const currentWidget = notebookTracker.currentWidget;
+    if (!currentWidget) {
+        console.warn('No active notebook found for restoration');
+        return false;
+    }
+
+    const notebook = currentWidget.content;
+    const notebookModel = notebook.model;
+    if (!notebookModel) {
+        console.warn('No notebook model found for restoration');
+        return false;
+    }
+
+    try {
+        console.log(`🔄 Restoring notebook from checkpoint: ${checkpoint.checkpointId}`);
+
+        // Clear all existing cells
+        const existingCellCount = notebookModel.cells.length;
+        for (let i = existingCellCount - 1; i >= 0; i--) {
+            notebookModel.sharedModel.deleteCell(i);
+        }
+
+        // Add cells from checkpoint
+        checkpoint.cells.forEach((cellCheckpoint, index) => {
+            const cellData = {
+                cell_type: cellCheckpoint.cellType,
+                source: cellCheckpoint.source,
+                metadata: cellCheckpoint.metadata || {}
+            } as {
+                cell_type: 'code' | 'markdown' | 'raw';
+                source: string;
+                metadata: Partial<ICellMetadata>;
+                outputs?: PartialJSONValue[];
+                execution_count?: number | null;
+            };
+
+            // Add code cell specific data
+            if (cellCheckpoint.cellType === 'code') {
+                cellData.outputs = cellCheckpoint.outputs || [];
+                cellData.execution_count = cellCheckpoint.executionCount;
```

Comment:
I don't think we're doing anything with these execution counts. We set them, but then they all get updated when we rerun th enotebook from top to bottom anyways, so I think we can delete?